File: chapter_0/bfs/min_deep/min_deep.rs
/*
 * @Author: Mengsen.Wang
 * @Date: 2021-01-21 19:56:06
 * @Last Modified by: Mengsen.Wang
 * @Last Modified time: 2021-01-24 22:47:34
 */

use std::cell::RefCell;
use std::rc::Rc;

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq, Clone)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

fn min_depth_bfs(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
    match root {
        None => 0,
        Some(root) => {
            let (mut depth, mut arr) = (1, vec![root]);
            'found: while !arr.is_empty() {
                let mut children = Vec::new();
                for node in arr {
                    if node.borrow().left.is_none() && node.borrow().right.is_none() {
                        break 'found;
                    }
                    if node.borrow().left.is_some() {
                        children.push(node.borrow().left.clone().unwrap());
                    }
                    if node.borrow().right.is_some() {
                        children.push(node.borrow().right.clone().unwrap());
                    }
                }
                arr = children;
                depth += 1;
            }
            depth
        }
    }
}

fn min_depth_dfs(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
    if root.is_none() {
        0
    } else {
        let (l, r) = (
            root.as_ref().unwrap().borrow().left.clone(),
            root.as_ref().unwrap().borrow().right.clone(),
        );
        match (l, r) {
            (None, None) => 1,
            (None, nd) | (nd, None) => 1 + min_depth_dfs(nd),
            (l, r) => 1 + std::cmp::min(min_depth_dfs(l), min_depth_dfs(r)),
        }
    }
}

fn main() {
    let mut root: TreeNode = TreeNode::new(3);
    let mut l1: TreeNode = TreeNode::new(2);
    let l2: TreeNode = TreeNode::new(1);
    let ll1: TreeNode = TreeNode::new(1);
    l1.left = Some(Rc::new(RefCell::new(ll1)));
    root.left = Some(Rc::new(RefCell::new(l1)));
    root.right = Some(Rc::new(RefCell::new(l2)));

    println!(
        "{}",
        min_depth_bfs(Some(Rc::new(RefCell::new(root.clone())))),
    );
    println!(
        "{}",
        min_depth_dfs(Some(Rc::new(RefCell::new(root.clone()))))
    );
}

File: cpp/min_deep/min_deep.cpp
/*
 * @Author: Mengsen.Wang
 * @Date: 2021-01-21 19:55:54
 * @Last Modified by: Mengsen.Wang
 * @Last Modified time: 2021-01-24 22:47:35
 */

#include <iostream>
#include <queue>

using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

int min_deep_bfs(TreeNode *root) {
  if (root == nullptr) return 0;
  queue<TreeNode *> que;
  que.push(root);
  int depth = 1;

  while (!que.empty()) {
    int size = que.size();
    for (int i = 0; i < size; ++i) {
      TreeNode *cul = que.front();
      if (cul->left == nullptr && cul->right == nullptr) return depth;
      if (cul->left != nullptr) que.push(cul->left);
      if (cul->right != nullptr) que.push(cul->right);
      que.pop();
    }
    ++depth;
  }
  return depth;
}

int min_deep_dfs(TreeNode *root) {
  if (root == nullptr) return 0;
  int left = min_deep_dfs(root->left);
  int right = min_deep_dfs(root->right);
  if (!left || !right) return left + right + 1;
  return min(left, right) + 1;
}

int main() {
  TreeNode root = TreeNode{3};
  TreeNode l1 = TreeNode{2};
  TreeNode l2 = TreeNode{1};
  TreeNode ll1 = TreeNode{1};
  l1.left = &ll1;
  root.left = &l1;
  root.right = &l2;
  std::cout << min_deep_bfs(&root) << std::endl;
  std::cout << min_deep_dfs(&root) << std::endl;
  return 0;
}